pennylane_lightning/src/simulator/OpToMemberFuncPtr.hpp

```diff
@@ -0,0 +1,444 @@
+                                const std::vector<size_t> &, bool);
+};
+} // namespace Internal
+
+/**
+ * @brief Convinient type alias for GateFuncPtr. See GateFuncPtr for details.
+ */
+template <class PrecisionT, class ParamT, size_t num_params>
+using GateFuncPtrT =
+    typename Internal::GateFuncPtr<PrecisionT, ParamT, num_params>::Type;
+
+template <class PrecisionT>
+using GeneratorFuncPtrT = typename Internal::GeneratorFuncPtr<PrecisionT>::Type;
+
+/**
+ * @defgroup Call gate operation with provided arguments
+ *
+ * @tparam PrecisionT Floating point type for the state-vector.
+ * @tparam ParamT Floating point type for the gate paramters.
+ * @param func Function pointer for the gate operation.
+ * @param num_qubits The number of qubits of the state-vector.
+ * @param wires Wires the gate applies to.
+ * @param inverse If true, we apply the inverse of the gate.
+ * @param params The list of gate paramters.
+ */
+/// @{
+/**
+ * @brief Overload for a gate operation without parameters
+ */
+template <class PrecisionT, class ParamT>
+inline void callGateOps(GateFuncPtrT<PrecisionT, ParamT, 0> func,
+                        std::complex<PrecisionT> *data, size_t num_qubits,
+                        const std::vector<size_t> &wires, bool inverse,
+                        [[maybe_unused]] const std::vector<ParamT> &params) {
+    assert(params.empty());
+    func(data, num_qubits, wires, inverse);
+}
+
+/**
+ * @brief Overload for a gate operation for a single paramter
+ */
+template <class PrecisionT, class ParamT>
+inline void callGateOps(GateFuncPtrT<PrecisionT, ParamT, 1> func,
+                        std::complex<PrecisionT> *data, size_t num_qubits,
+                        const std::vector<size_t> &wires, bool inverse,
+                        const std::vector<ParamT> &params) {
+    assert(params.size() == 1);
+    func(data, num_qubits, wires, inverse, params[0]);
+}
+
+/**
+ * @brief Overload for a gate operation for three paramters
+ */
+template <class PrecisionT, class ParamT>
+inline void callGateOps(GateFuncPtrT<PrecisionT, ParamT, 3> func,
+                        std::complex<PrecisionT> *data, size_t num_qubits,
+                        const std::vector<size_t> &wires, bool inverse,
+                        const std::vector<ParamT> &params) {
+    assert(params.size() == 3);
+    func(data, num_qubits, wires, inverse, params[0], params[1], params[2]);
+}
+/// @}
+/**
+ * @brief Call a generator operation.
+ *
+ * @tparam PrecisionT Floating point type for the state-vector.
+ * @return Scaling factor
+ */
+template <class PrecisionT>
+inline PrecisionT callGeneratorOps(GeneratorFuncPtrT<PrecisionT> func,
+                                   std::complex<PrecisionT> *data,
+                                   size_t num_qubits,
+                                   const std::vector<size_t> &wires, bool adj) {
+    return func(data, num_qubits, wires, adj);
+}
+} // namespace Pennylane
```

Comment:
I'd love to get an understanding of this class at some point too! Great work mapping it out.

Reply:
This file somehow happens to be the static version of the switch-case. Besides of that, I would there there is nothing much special :wink: